csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/NlpDispatchBot.cs

```diff
@@ -0,0 +1,182 @@
+    /// For each interaction from the user, an instance of this class is called.
+    /// This is a Transient lifetime service.  Transient lifetime services are created
+    /// each time they're requested. For each Activity received, a new instance of this
+    /// class is created. Objects that are expensive to construct, or have a lifetime
+    /// beyond the single Turn, should be carefully managed.
+    /// </summary>
+    public class NlpDispatchBot : IBot
+    {
+        /// <summary>
+        /// Key in the Bot config (.bot file) for the Home Automation Luis instance.
+        /// </summary>
+        public static readonly string HomeAutomationLuisKey = "homeautomation.luis";
+
+        /// <summary>
+        /// Key in the Bot config (.bot file) for the Weather Luis instance.
+        /// </summary>
+        public static readonly string WeatherLuisKey = "weather.luis";
+
+        /// <summary>
+        /// Key in the Bot config (.bot file) for the Dispatch.
+        /// </summary>
+        public static readonly string DispatchKey = "bot-dispatch";
+
+        /// <summary>
+        /// Key in the Bot config (.bot file) for the QnaMaker instance.
+        /// In the .bot file, multiple instances of QnaMaker can be configured.
+        /// </summary>
+        public static readonly string QnAMakerKey = "sample.qna";
+
+        /// <summary>
+        /// Services configured from the ".bot" file.
+        /// </summary>
+        private readonly BotServices _services;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NlpDispatchBot"/> class.
+        /// </summary>
+        /// <param name="services">Services configured from the ".bot" file.</param>
+        public NlpDispatchBot(BotServices services)
+        {
+            _services = services ?? throw new System.ArgumentNullException(nameof(services));
+
+            if (!_services.QnAServices.ContainsKey(QnAMakerKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a QnA service named '{DispatchKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(HomeAutomationLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{HomeAutomationLuisKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(WeatherLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{WeatherLuisKey}'.");
+            }
+        }
+
+        /// <summary>
+        /// Every Conversation turn for our NLP Dispatch Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response, with no stateful conversation.
```

Comment:
>with no stateful conver [](start = 34, length = 23)

If this is the case, why is state setup in the main bot? Memory Storage is configured and it probably doesn't need to be.

Reply:
Where is state set up? I don't think I set up memory storage, at least not explicitly. Unless I'm missing something :) (even search for "memory" in the solution yields zero results)

Reply:
You're right, of course. My mistake. :) 

---
In reply to: [215322806](https://github.com/Microsoft/BotBuilder-Samples/pull/319#discussion_r215322806) [](ancestors = 215322806)